In [1]:
# Import libraries
import pandas as pd

# Read the CSV and Perform Basic Data Cleaning

In [2]:
# Read CSV file 
df = pd.read_csv("exoplanet_data.csv")

# Drop the null columns where all values are null
df = df.dropna(axis='columns', how='all')

# Drop the null rows
df = df.dropna()
df.head(10)

koi_disposition  koi_fpflag_nt  koi_fpflag_ss  koi_fpflag_co  koi_fpflag_ec  \
0       CONFIRMED              0              0              0              0   
1  FALSE POSITIVE              0              1              0              0   
2  FALSE POSITIVE              0              1              0              0   
3       CONFIRMED              0              0              0              0   
4       CONFIRMED              0              0              0              0   
5       CONFIRMED              0              0              0              0   
6       CONFIRMED              0              0              0              0   
7       CONFIRMED              0              0              0              0   
8       CONFIRMED              0              1              0              0   
9       CONFIRMED              0              0              0              0   

   koi_period  koi_period_err1  koi_period_err2  koi_time0bk  \
0   54.418383     2.479000e-04    -2.479000e-04   162.513840   
1   19.899140     1.490000e-05    -1.490000e-05   175.850252   
2    1.736952     2.630000e-07    -2.630000e-07   170.307565   
3    2.525592     3.760000e-06    -3.760000e-06   171.595550   
4    4.134435     1.050000e-05    -1.050000e-05   172.979370   
5    2.566589     1.780000e-05    -1.780000e-05   179.554370   
6   16.068647     1.090000e-05    -1.090000e-05   173.621937   
7    2.470613     2.700000e-08    -2.700000e-08   122.763305   
8    2.204735     4.300000e-08    -4.300000e-08   121.358542   
9    3.522498     1.980000e-07    -1.980000e-07   121.119423   

   koi_time0bk_err1  ...  koi_steff_err2  koi_slogg  koi_slogg_err1  \
0          0.003520  ...             -81      4.467           0.064   
1          0.000581  ...            -176      4.544           0.044   
2          0.000115  ...            -174      4.564           0.053   
3          0.001130  ...            -211      4.438           0.070   
4          0.001900  ...            -232      4.486           0.054   
5          0.004610  ...            -232      4.486           0.054   
6          0.000517  ...             -83      4.485           0.083   
7          0.000009  ...             -78      4.457           0.024   
8          0.000016  ...             -89      4.019           0.033   
9          0.000047  ...            -137      4.169           0.055   

   koi_slogg_err2  koi_srad  koi_srad_err1  koi_srad_err2         ra  \
0          -0.096     0.927          0.105         -0.061  291.93423   
1          -0.176     0.868          0.233         -0.078  297.00482   
2          -0.168     0.791          0.201         -0.067  285.53461   
3          -0.210     1.046          0.334         -0.133  288.75488   
4          -0.229     0.972          0.315         -0.105  296.28613   
5          -0.229     0.972          0.315         -0.105  296.28613   
6          -0.028     0.848          0.033         -0.072  286.99948   
7          -0.024     0.964          0.038         -0.038  286.80847   
8          -0.027     1.952          0.099         -0.110  292.24728   
9          -0.045     1.451          0.110         -0.110  281.28812   

         dec  koi_kepmag  
0  48.141651      15.347  
1  48.134129      15.436  
2  48.285210      15.597  
3  48.226200      15.509  
4  48.224670      15.714  
5  48.224670      15.714  
6  48.375790      15.841  
7  49.316399      11.338  
8  47.969521      10.463  
9  42.451080      13.563  

[10 rows x 41 columns]

# Select features (columns)

In [3]:
# Set features
# selected_features = df[['names', 'of', 'selected', 'features', 'here']]
selected_features = df

# Create a Train Test Split
Use koi_disposition for the y values

In [5]:
# Assign x and y values
X = df.drop("koi_disposition", axis=1)
y = df["koi_disposition"]

print(X.shape, y.shape)

(6991, 40) (6991,)


In [6]:
# Import Dependencies
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import LabelEncoder, MinMaxScaler
from tensorflow.keras.utils import to_categorical

# Split data into training and testing
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=115, stratify=y)

In [7]:
X_train.head()

koi_fpflag_nt  koi_fpflag_ss  koi_fpflag_co  koi_fpflag_ec  koi_period  \
2700              0              0              0              0   31.804760   
21                0              0              0              0    4.280964   
5040              0              1              1              0    0.829019   
2795              0              0              1              0    1.387838   
1765              0              0              0              0   14.383227   

      koi_period_err1  koi_period_err2  koi_time0bk  koi_time0bk_err1  \
2700         0.000091        -0.000091    190.25413           0.00224   
21           0.000006        -0.000006    171.89659           0.00115   
5040         0.000002        -0.000002    131.46667           0.00401   
2795         0.000006        -0.000006    132.11351           0.00318   
1765         0.000086        -0.000086    140.57648           0.00530   

      koi_time0bk_err2  ...  koi_steff_err2  koi_slogg  koi_slogg_err1  \
2700          -0.00224  ...            -128      4.235           0.143   
21            -0.00115  ...            -169      4.564           0.032   
5040          -0.00401  ...            -255      4.271           0.124   
2795          -0.00318  ...            -219      4.442           0.054   
1765          -0.00530  ...            -135      4.249           0.137   

      koi_slogg_err2  koi_srad  koi_srad_err1  koi_srad_err2         ra  \
2700          -0.117     1.356          0.212         -0.259  287.62082   
21            -0.168     0.831          0.207         -0.069  294.26581   
5040          -0.186     1.328          0.408         -0.220  296.03485   
2795          -0.216     1.044          0.335         -0.112  295.64212   
1765          -0.112     1.291          0.216         -0.216  290.08276   

            dec  koi_kepmag  
2700  43.141647      13.178  
21    49.314091      15.356  
5040  45.490822      14.661  
2795  48.773071      15.167  
1765  50.863480      13.312  

[5 rows x 40 columns]

# Pre-processing
Scale the data using the MinMaxScaler and perform some feature selection

In [8]:
# Scale the data
X_scaler = MinMaxScaler().fit(X_train)
X_train_scaled = X_scaler.transform(X_train)
X_test_scaled = X_scaler.transform(X_test)

# Train Model - SVM
Use Support Vector Machine (SVM) model

In [9]:
# Create the SVC Model
from sklearn.svm import SVC 
model = SVC(kernel='linear')
model.fit(X_train_scaled, y_train)

SVC(kernel='linear')

In [10]:
print(f"Training Data Score: {model.score(X_train_scaled, y_train)}")
print(f"Testing Data Score: {model.score(X_test_scaled, y_test)}")

Training Data Score: 0.8498950982262063
Testing Data Score: 0.8329519450800915


# Hyperparameter Tuning
Use GridSearchCV to tune the SVM model's parameters

In [11]:
# Create the GridSearchCV model
from sklearn.model_selection import GridSearchCV
param_grid = {'C': [1, 5, 10],
              'gamma': [0.0001, 0.001, 0.01]}
grid = GridSearchCV(model, param_grid, verbose=3)

In [12]:
# Fit the SVM model using the grid search estimator. 
# This will take the SVM model and try each combination of parameters
grid.fit(X_train_scaled, y_train)

Fitting 5 folds for each of 9 candidates, totalling 45 fits
[CV] C=1, gamma=0.0001 ...............................................


[Parallel(n_jobs=1)]: Using backend SequentialBackend with 1 concurrent workers.


[CV] ................... C=1, gamma=0.0001, score=0.842, total=   0.3s
[CV] C=1, gamma=0.0001 ...............................................


[Parallel(n_jobs=1)]: Done   1 out of   1 | elapsed:    0.3s remaining:    0.0s


[CV] ................... C=1, gamma=0.0001, score=0.860, total=   0.3s
[CV] C=1, gamma=0.0001 ...............................................


[Parallel(n_jobs=1)]: Done   2 out of   2 | elapsed:    0.6s remaining:    0.0s


[CV] ................... C=1, gamma=0.0001, score=0.841, total=   0.3s
[CV] C=1, gamma=0.0001 ...............................................
[CV] ................... C=1, gamma=0.0001, score=0.843, total=   0.3s
[CV] C=1, gamma=0.0001 ...............................................
[CV] ................... C=1, gamma=0.0001, score=0.844, total=   0.3s
[CV] C=1, gamma=0.001 ................................................
[CV] .................... C=1, gamma=0.001, score=0.842, total=   0.3s
[CV] C=1, gamma=0.001 ................................................
[CV] .................... C=1, gamma=0.001, score=0.860, total=   0.3s
[CV] C=1, gamma=0.001 ................................................
[CV] .................... C=1, gamma=0.001, score=0.841, total=   0.3s
[CV] C=1, gamma=0.001 ................................................
[CV] .................... C=1, gamma=0.001, score=0.843, total=   0.3s
[CV] C=1, gamma=0.001 ................................................
[CV] .

[Parallel(n_jobs=1)]: Done  45 out of  45 | elapsed:   14.9s finished


GridSearchCV(estimator=SVC(kernel='linear'),
             param_grid={'C': [1, 5, 10], 'gamma': [0.0001, 0.001, 0.01]},
             verbose=3)

In [13]:
# List the best parameters for this dataset
print(f"Best Parameters: {grid.best_params_}")

# List the best score
print(f"Best Score: {grid.best_score_}")

Best Parameters: {'C': 10, 'gamma': 0.0001}
Best Score: 0.8733555767397521


In [14]:
print(f"Training Grid Score: {grid.score(X_train_scaled, y_train)}")
print(f"Testing Grid Score: {grid.score(X_test_scaled, y_test)}")

Training Grid Score: 0.8775510204081632
Testing Grid Score: 0.8655606407322655


In [16]:
 #Make predictions with the hypertuned model
predictions = grid.predict(X_test_scaled)

In [17]:
# Calculate classification report
from sklearn.metrics import classification_report
print(classification_report(y_test, predictions))

                precision    recall  f1-score   support

     CANDIDATE       0.83      0.58      0.68       422
     CONFIRMED       0.70      0.88      0.78       450
FALSE POSITIVE       0.99      1.00      0.99       876

      accuracy                           0.87      1748
     macro avg       0.84      0.82      0.82      1748
  weighted avg       0.87      0.87      0.86      1748



# Save the Model

In [18]:
# Save the model 
import joblib
filename = 'MYCROFTXTREEM_SVM.sav'
joblib.dump(model, filename)

['MYCROFTXTREEM_SVM.sav']

Conclusion





The f1-score of FALSE POSITIVE for SVM model is closer to 1 (0.99) meaning it can predict FALSE POSITIVE well and it's reliable.

The Grid score from the hyperparameter tuning boasts the SVM score higher Training Data Score: 0.8498950982262063 changes to 0.8775510204081632 Testing Data Score: 0.8329519450800915 changes to 0.8655606407322655 Best Score: 0.8706847224871257

However, Random Forest model's best score of (0.8987221563248168) seems better than SVM model (0.8706847224871257) when comparing the scores.

